unixfs/mod/dagmodifier.go

```diff
@@ -238,43 +232,87 @@ func (dm *DagModifier) Sync() error {
 // returns the new key of the passed in node and whether or not all the data in the reader
 // has been consumed.
 func (dm *DagModifier) modifyDag(n node.Node, offset uint64, data io.Reader) (*cid.Cid, bool, error) {
-	node, ok := n.(*mdag.ProtoNode)
-	if !ok {
-		return nil, false, ErrNoRawYet
-	}
+	// If we've reached a leaf node.
+	if len(n.Links()) == 0 {
+		switch nd0 := n.(type) {
+		case *mdag.ProtoNode:
+			f, err := ft.FromBytes(nd0.Data())
+			if err != nil {
+				return nil, false, err
+			}
 
-	f, err := ft.FromBytes(node.Data())
-	if err != nil {
-		return nil, false, err
-	}
+			n, err := data.Read(f.Data[offset:])
+			if err != nil && err != io.EOF {
+				return nil, false, err
+			}
 
-	// If we've reached a leaf node.
-	if len(node.Links()) == 0 {
-		n, err := data.Read(f.Data[offset:])
-		if err != nil && err != io.EOF {
-			return nil, false, err
-		}
+			// Update newly written node..
+			b, err := proto.Marshal(f)
+			if err != nil {
+				return nil, false, err
+			}
 
-		// Update newly written node..
-		b, err := proto.Marshal(f)
-		if err != nil {
-			return nil, false, err
-		}
+			nd := new(mdag.ProtoNode)
+			nd.SetData(b)
+			nd.SetPrefix(&nd0.Prefix)
+			k, err := dm.dagserv.Add(nd)
+			if err != nil {
+				return nil, false, err
+			}
 
-		nd := new(mdag.ProtoNode)
-		nd.SetData(b)
-		k, err := dm.dagserv.Add(nd)
-		if err != nil {
-			return nil, false, err
-		}
+			// Hey look! we're done!
+			var done bool
+			if n < len(f.Data[offset:]) {
+				done = true
+			}
 
-		// Hey look! we're done!
-		var done bool
-		if n < len(f.Data[offset:]) {
-			done = true
+			return k, done, nil
+		case *mdag.RawNode:
+			origData := nd0.RawData()
+			bytes := make([]byte, len(origData))
```

Comment:
I think its probably okay to *not* make a copy of the data here, but i guess we can optimize this later (and as is, is safe)